content/blog/2018-06-07-when-to-use-derived-state.md

```diff
@@ -0,0 +1,384 @@
+
+## When should I avoid derived state?
+
+The terms ["controlled"](/docs/forms.html#controlled-components) and ["uncontrolled"](/docs/uncontrolled-components.html) are often used to refer to form components, but they can also be used to describe where a component's state lives. Data that is passed in as props can be thought of as **controlled** and one that exists only in internal state can be thought of as **uncontrolled**.
+
+The most common mistake with derived state is mixing these two. In other words, when props control a value in state that is also updated by `setState` calls. This may sound similar to the [external data loading example](/blog/2018/03/27/update-on-async-rendering.html#fetching-external-data-when-props-change) mentioned above, but it's different in a few important ways.
+
+In the loading example, there is a clear source of truth for both the "source" prop and the "loading" state. When the source prop changes, the loading state should **always** be overridden. Conversely the state should only be overridden when the prop **changes** and should otherwise be managed by the component.
+
+Problems arise when any of these constraints are changed. This typically comes in two forms. Let's take a look at both.
+
+### Anti-pattern: Props always override state
+
+A common misconception about both `getDerivedStateFromProps` and `componentWillReceiveProps` is that they are only called when props "change". However, any time a parent component re-renders, the child component still receives the new props. Because of this, it is (and has always been) unsafe to _unconditionally_ override state values using either of these lifecycles.
+
+Before React 16.4, it was difficult to reproduce such bugs consistently because they depend on the timing of parent re-rendering. To expose them consistently, we made `getDerivedStateFromProps` in React 16.4 fire for renders caused by `setState` or `forceUpdate` too. Let’s consider an example to demonstrate the problem.
+
+Here is a `EmailInput` component that accepts an email prop:
+```js
+class EmailInput extends Component {
+  state = {
+    email: this.props.email
+  };
+
+  static getDerivedStateFromProps(props, state) {
+    // This is bad! Do not do this!
+    if (props.email) {
+      return { email: props.email };
+    }
+    return null;
+  }
+
+  render() {
+    return <input onChange={this.handleChange} value={this.state.email} />;
+  }
+
+  handleChange = event => {
+    this.setState({ email: event.target.value });
+  };
+}
+```
+
+At first, this component might look okay. State is initialized to the value specified by props and updated when we type into the `<input>`. But once our component re-renders– either because it called `setState` or because its parent re-rendered– anything we've typed into the `<input>` will be lost! (See [this demo for an example](codesandbox://when-to-use-derived-state/derived-state-anti-pattern).)
```

Comment:
The demo code looks different than the snippet above.

Reply:
Yeah. I intended the demo to show the general problem, not really to exactly mirror the code so far. (I started mirroring the code, but then I felt that writing a validate function was weird, and using it kind of distracted from the point I was trying to make...)

Reply:
It can just be an empty function, no? That would still show the point.

My concern is that it's confusing to have the gDSFP implementation itself be different when the post is highlighting antipatterns in gDSFP. This makes pattern-matching style of learning harder.

Reply:
Note I haven't looked at latest edit yet

Reply:
The confusion I was concerned about was _how_ the validation function is used. Does it sit between the change handler and the `setState` call? Yeah, probably- but now we're introducing a new kind of semi-controlled behavior into the example which I was worried might distract.

I _think_ I kind of hand waved over this in the latest Code Sandbox, but feedback is welcome.